File: node/test/teacher.test.js
const app = require('../server/app'),
  chai = require('chai'),
  request = require('supertest');

const expect = chai.expect;
const util = require('./util');


describe('Teachers', () => {
  // It works by itself, so remember to run this manually
  it.skip('Should Get An Empty List Of Teachers', async () => {
    const res = await request(app)
      .get("/teachers");

    expect(res.statusCode).to.equal(200);
    expect(res.body).to.be.an('array');
    expect(res.body).to.be.empty;
  });

  it('Should Get List Of Students', async () => {
    const testTeacher = await util.createTestTeacher(app);

    const res = await request(app)
      .get("/teachers");

    expect(res.statusCode).to.equal(200);
    expect(res.body).to.be.an('array');
    expect(res.body).to.deep.include(testTeacher);
  });

  it('Should Add A Student To A Teacher', async () => {
    const testTeacher = await util.createTestTeacher(app);
    const testStudent = await util.createTestStudent(app);

    const putBody = { studentId: testStudent.id };
    const res = await request(app)
      .put(`/teachers/${testTeacher.id}`)
      .send(putBody);

    expect(res.statusCode).to.equal(200);
    const updatedTeacher = res.body;
    expect(updatedTeacher.students).to.deep.include(testStudent);
  });

});

File: dotnet/InterviewTest.Tests/TeacherTest.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Nancy;
using Nancy.Testing;
using Xunit;

namespace InterviewTest.Tests
{
    public class TeacherTest : IDisposable
    {
        
        [Fact]
        public async Task Should_GetAnEmptyListOfTeachersAsync()
        {
            var bootstrapper = new DefaultNancyBootstrapper();
            var browser = new Browser(bootstrapper);

            var result = await browser.Get("/teachers", with => with.HttpRequest());

            Assert.Equal(HttpStatusCode.OK, result.StatusCode);
            Assert.Equal("[]", result.Body.AsString());
        }

        [Fact]
        public async Task Should_GetListOfTeachersAsync()
        {
            var bootstrapper = new DefaultNancyBootstrapper();
            var browser = new Browser(bootstrapper);
            var testTeacher = await browser.CreateTestTeacherAsync();

            var result = await browser.Get("/teachers", with => with.HttpRequest());

            Assert.Equal(HttpStatusCode.OK, result.StatusCode);
            var teacherList = result.Body.DeserializeJson<Teacher[]>();
            Assert.Single(teacherList, testTeacher);
        }

        [Fact(Skip = "It works by itself, so remember to run this manually")]
        public async Task Should_AddAStudentToATeacherAsync()
        {
            var bootstrapper = new DefaultNancyBootstrapper();
            var browser = new Browser(bootstrapper);
            var testTeacher = await browser.CreateTestTeacherAsync();
            var testStudent = await browser.CreateTestStudentAsync();

            var putBody = new {studentId = testStudent.Id};
            var result = await browser.Put($"/teachers/{testTeacher.Id}",
                with =>
                {
                    with.HttpRequest();
                    with.JsonBody(putBody);
                });

            Assert.Equal(HttpStatusCode.OK, result.StatusCode);
            var updatedTeacher = result.Body.DeserializeJson<Teacher>();
            Assert.Single(updatedTeacher.Students, testStudent);
        }

        public void Dispose()
        {
            StudentCollection.GetInstance().Clear();
            TeacherCollection.GetInstance().Clear();
        }
    }
}